File: mySolutions/127. Word Ladder/127. Word Ladder.4.py
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
      if endWord not in wordList:
        return 0
      wordList.remove(endWord)

      wordListSet = set(wordList)
      leftSet = {beginWord}
      rightSet = {endWord}
      visited = set()
      length = 0

      while len(leftSet) > 0 and len(rightSet) > 0:
        # print("leftSet", leftSet, "rightSet", rightSet)
        nextSet = set()
        length += 1
        
        for word in leftSet:
          # try all the conversion
          for i in range(len(word)):
            nextWordArr = list(word)

            for j in range(26):
              ch = chr(ord('a') + j)
              if ch == nextWordArr[i]:
                continue
              
              nextWordArr[i] = ch
              nextWord = ''.join(nextWordArr)
              if nextWord in rightSet:
                return length + 1

              if (nextWord in wordListSet) and (nextWord not in visited):
                nextSet.add(nextWord)
                visited.add(nextWord)
          
        leftSet = nextSet

        if len(leftSet) > len(nextSet):
          temp = leftSet
          leftSet = rightSet
          rightSet = temp

      return 0

File: mySolutions/127. Word Ladder/127. Word Ladder.4.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        
        Set<String> unused = new HashSet<>(wordList);
        Queue<String> queue = new LinkedList<>();
        
        queue.offer(beginWord);
        int depth = 1;
        
        while (queue.size() != 0) {
            int size = queue.size();
            depth += 1;
            
            for (int i = 0; i < size; i++) {
                String curr = queue.poll();
                // System.out.printf("curr: %s\n", curr);
                for (int j = 0; j < curr.length(); j++) {
                    char[] chars = curr.toCharArray();
                    for (int k = 0; k < 26; k++) {
                        chars[j] = (char)('a' + k);
                        String nw = String.valueOf(chars);
                        if (unused.contains(nw)) {
                            queue.offer(nw);
                            unused.remove(nw);
                            if (nw.equals(endWord)) {
                                return depth;
                            }
                        }
                    }
                }
            }
        }
        
        return 0;
    }
}